### DISCUSSION TITLE
Deprecated `request.page` in middleware

### DISCUSSION THREAD

**Discussion by itsikcircle** [vercel/next.js#38159] [Help] [answered]
I've been using `request.page` in the middleware to get the page url with slug, but after deprecating it in `12.2.0` it's not possible anymore to know the page url with slug. (https://github.com/vercel/next.js/pull/37349)

For example, I have pages like:
- /page/[id]
- /page/[...slug]

But not I can only get
- /page/1234
- /page/some/thing
- /page/some/url

It's an issue for me since I have a list of all the urls I want to handle with `slug`, but in case of absolute urls with id, it makes it hard to extract and match urls.

Any idea how I can solve it? as the suggestion to use `URLPattern` is not working well for my use case.

**Comment by icyJoseph** [comment]
mmm perhaps a bit dirty, but... 

```ts
const PATTERNS = [
  [new URLPattern({ pathname: "/:id" }), ({ pathname }) => pathname.groups],
  [new URLPattern({ pathname: "/:slug+" }), ({ pathname }) => pathname.groups],
];

// and then

for (const [pattern, handler] of PATTERNS) {
  // using window.location.origin so that you can see this in the browser console
  const patternResult = pattern.exec(window.location.origin + "/something");
  if (patternResult !== null && "pathname" in patternResult) {
    console.log(pattern.pathname, handler(patternResult));
    break;
  }
}
for (const [pattern, handler] of PATTERNS) {
  // using window.location.origin so that you can see this in the browser console
  const patternResult = pattern.exec(window.location.origin + "/some/thing");
  if (patternResult !== null && "pathname" in patternResult) {
    console.log(pattern.pathname, handler(patternResult));
    break;
  }
}
```

I'd imagine you maybe tried this already? It is kinda fragile to the other of PATTERNS. You also have to split the slug if you match it... not precisely straightforward as it used to be.

Or is it that there's `/pages/something` that you'd want to match as `[...slug]`? But that's against the rules for dynamic routes:

```markdown
- pages/post/create.js - Will match /post/create
- pages/post/[pid].js - Will match /post/1, /post/abc, etc. But not /post/create
- pages/post/[...slug].js - Will match /post/1/2, /post/a/b/c, etc. But not /post/create, /post/abc
```

Could you expand a bit more on your problem? I am about to upgrade a few apps to 12.2.0, and one of them has this... IIRC

**Reply by itsikcircle** [reply]
I've tried it, no success so far.

I have an object with list of all valid routes, while each route have some sort of configuration object, for example
```ts
const routes = {
  'pages/posts': {
    example: true,
    someKey: 'keyName'
  },
  'pages/posts/[pid]': {
    example: true,
    someKey: 'keyName'
  }
  'pages/posts/[pid]/[comment]': {
    example: true,
    someKey: 'keyName'
  }
}
```
With `request.page.name` I will get the exact match of the dynamic route, so I can find the configuration of the route in the list.

With `12.2.0` it's not available anymore and there is no way to get the dynamic route, the only options is `request.url` or `request.nextUrl.pathname` which end up with the full url.

So now I have have this kind of urls
```
/pages/posts
/pages/posts/123
/pages/posts/123/789
```
- Using `url.startsWith` won't work as `/pages/posts` will match all 3 urls
- Using exact match won't work as `/pages/posts/123` won't match `pages/posts/[pid]` 

Keep in mind that some dynamic urls may have different name other than `pid`, so using regex to cover all use cases become a real challenge.

I've been trying to dig into the source code the find out how next know the dynamic route matching the url, but that require access to the pages list which is not available as well.

I'll add to that, there is no way to know which part of the url is the params. 
For example in the url `/pages/posts/123` the `123` is `id` param, if I had this information it will be possible to replace `/123` with `/[id]` and have the dynamic route.

**Comment by itsikcircle** [comment, marked as answer]
I've found a solution
```ts
import { getRouteRegex } from 'next/dist/shared/lib/router/utils/route-regex'

const route = getRouteRegex('/page/posts/[pid]/[comment]')

if (route.re.test('/page/posts/123/789')) {
  // matching route...
}
```

**Reply by bryanprimus** [reply]
Is this work as expected? What about the case when you have static routes alongside dynamic routes. 

page/[slug].tsx, page/create.tsx

**Reply by icyJoseph** [reply]
I think you'd need to test first see if the pathname is `/page/create` and if not then try to match it to the regex... it starts to depend on the order and, it expects your code to keep track of the file structure, but I guess it is the best we can do for now.

**Comment by bennettdams** [comment]
## Workaround & feature request

Regarding the [proposed solution here](https://github.com/vercel/next.js/discussions/38159#discussioncomment-3056398): I don't think using internal utils from the Next.js should be considered anything but a hack. Also, you'd need to define valid routes upfront, so it is easy to miss a newly created route.

I create a feature request to bring back the functionality of identifying a route request: https://github.com/vercel/next.js/discussions/38296

I also wrote down another workaround in that discussion.

**Reply by bryanprimus** [reply]
is there any temporary workaround for this? can't upgrade to v12.2 because of this

**Reply by bennettdams** [reply]
@bryantobing12 I wrote down a workaround in the linked discussion.
